src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/Huffman.cs

```diff
@@ -308,119 +284,152 @@ public static (uint encoded, int bitLength) Encode(int data)
+                        // if our bitOffset is zero then all of bits in this byte are set
+                        // "A padding strictly longer than 7 bits MUST be treated as a decoding error."
+                        // https://httpwg.org/specs/rfc7541.html#rfc.section.5.2
+                        if (bitOffset == 0)
+                            throw new HuffmanDecodingException();
 
-                // Modulo 8 since we only care about how many bits were decoded in the last byte that we processed.
-                lastDecodedBits %= 8;
+                        // otherwise (padding less than 8), just break out of the main loop because we are done
+                        break;
+                    }
+                }
             }
 
-            return j;
+            return destinationIndex;    // destinationIndex is incremented each time a value is decoded, so it also represents total decoded byte count
         }
 
-        /// <summary>
-        /// Decodes a single symbol from a 32-bit word.
-        /// </summary>
-        /// <param name="data">A 32-bit word containing a Huffman encoded symbol.</param>
-        /// <param name="validBits">
-        /// The number of bits in <paramref name="data"/> that may contain an encoded symbol.
-        /// This is not the exact number of bits that encode the symbol. Instead, it prevents
-        /// decoding the lower bits of <paramref name="data"/> if they don't contain any
-        /// encoded data.
-        /// </param>
-        /// <param name="decodedBits">The number of bits decoded from <paramref name="data"/>.</param>
-        /// <returns>The decoded symbol.</returns>
-        public static int Decode(uint data, int validBits, out int decodedBits)
+        private static short[,] BuildDecodingArray()
         {
-            // The code below implements the decoding logic for a canonical Huffman code.
-            //
-            // To decode a symbol, we scan the decoding table, which is sorted by ascending symbol bit length.
-            // For each bit length b, we determine the maximum b-bit encoded value, plus one (that is codeMax).
-            // This is done with the following logic:
-            //
-            // if we're at the first entry in the table,
-            //    codeMax = the # of symbols encoded in b bits
-            // else,
-            //    left-shift codeMax by the difference between b and the previous entry's bit length,
-            //    then increment codeMax by the # of symbols encoded in b bits
-            //
-            // Next, we look at the value v encoded in the highest b bits of data. If v is less than codeMax,
-            // those bits correspond to a Huffman encoded symbol. We find the corresponding decoded
-            // symbol in the list of values associated with bit length b in the decoding table by indexing it
-            // with codeMax - v.
-
-            int codeMax = 0;
+            var array = new short[15, 256];
```

Comment:
~ 7.5k

Reply:
This is my biggest concern with the array method. We are allocating a non-trivial chuck of memory that does not get reclaimed.

Reply:
Note that initialization of `s_encodingTable` is 9kB+ of native code and another 2kB on GC heap for the table itself. If you wanted to make this more efficient, the place to focus on is the `s_encodingTable` initialization.